### ISSUE TITLE
Wrong type inference with `is_a?` on generic types with `upper` bound

### ISSUE THREAD

**Issue by Morriar** [sorbet/sorbet#1604]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20extend%20T%3A%3AGeneric%0A%20%20Elem%20%3D%20type_template(upper%3A%20String)%0Aend%0A%0Aclass%20B%3B%20end%0A%0Aclass%20Foo%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7B%20params(a%3A%20T.any(A%2C%20B)).void%20%7D%0A%20%20def%20error(a)%0A%20%20%20%20if%20a.is_a%3F(A)%0A%20%20%20%20%20%20T.reveal_type(a)%20%23%20Should%20be%20%60A%60%20or%20%60A%5BT.untyped%5D%60%0A%20%20%20%20end%0A%20%20end%0Aend)

```ruby
# typed: true

class A
  extend T::Generic
  Elem = type_template(upper: String)
end

class B; end

class Foo
  extend T::Sig

  sig { params(a: T.any(A, B)).void }
  def error(a)
    if a.is_a?(A)
      T.reveal_type(a) # Should be `A` or `A[T.untyped]`
    end
  end
end
```

#### Observed output

```
editor.rb:16: Revealed type: T.any(A, B) https://srb.help/7014
    16 |      T.reveal_type(a) # Should be `A` or `A[T.untyped]`
              ^^^^^^^^^^^^^^^^
  From:
    editor.rb:14:
    14 |  def error(a)
                    ^
Errors: 1
```

#### Expected behavior

Since Sorbet features [flow typing](https://sorbet.org/docs/flow-sensitive) I would expect the revealed type to be `A` or `A[T.untyped]` maybe even `A[String]` since we know the upper bound.

Changing the generic bound from `upper` to `fixed` actually gives the result I was expecting: `A`

https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20extend%20T%3A%3AGeneric%0A%20%20Elem%20%3D%20type_template(fixed%3A%20String)%0Aend%0A%0Aclass%20B%3B%20end%0A%0Aclass%20Foo%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7B%20params(a%3A%20T.any(A%2C%20B)).void%20%7D%0A%20%20def%20error(a)%0A%20%20%20%20if%20a.is_a%3F(A)%0A%20%20%20%20%20%20T.reveal_type(a)%20%23%20Should%20be%20%60A%60%20or%20%60A%5BT.untyped%5D%60%0A%20%20%20%20end%0A%20%20end%0Aend

**Comment by elliottt** [comment]
`fixed` type variables are a [special case](https://github.com/sorbet/sorbet/blob/703e91a3ea002ffd7c9a700396449658f8a04209/core/types/types.cc#L849-L857) in the code that handles types given to `is_a?`, which explains why you get `A` when you change it.

This is a known problem, but I don't believe we have an existing issue. Thanks for the bug report!

**Comment by elliottt** [comment]
This should be fixed with #1617.
